### PR TITLE
[OptionsResolver] Passing Options argument to deprecation closure

### PR DESCRIPTION
**Pull request by yceruto**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | https://github.com/symfony/symfony/pull/28721#discussion_r222686221
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/pull/10439

As spotted here https://github.com/symfony/symfony/pull/28721, we sometimes need more advanced cases, where the deprecation of the value depends on another option:
```php
$resolver->setDeprecated('date_format', function (Options $options, $dateFormat) {
    if (null !== $options['date_format'] && 'single_text' === $options['widget']) {
         return sprintf('Using the "date_format" option of the %s when the "widget" option is set to "single_text" is deprecated since Symfony 4.2.', self::class);
    }

    return '';
});
```
There is still a decision to make:
> We're in time to change the arguments position (Options $options, $value) to be consistent with other closure signatures.

WDYT?

### CODE REVIEW

**Review by ogizanagi — APPROVED** [review]
👍

**Comment by xabbuh** [comment]
> We're in time to change the arguments position (Options $options, $value) to be consistent with other closure signatures.

Yes, let's do that. Otherwise, it will become hard to remember when arguments are passed in which order.

File: src/Symfony/Component/OptionsResolver/OptionsResolver.php [hunk 1]
```diff
@@ -860,8 +860,20 @@ public function offsetGet($option)
...
+            if ($deprecationMessage instanceof \Closure) {
+                // If the closure is already being called, we have a cyclic dependency
+                if (isset($this->calling[$option])) {
+                    throw new OptionDefinitionException(sprintf('The options "%s" have a cyclic dependency.', implode('", "', array_keys($this->calling))));
+                }
+
+                $this->calling[$option] = true;
+                try {
+                    if (!\is_string($deprecationMessage = $deprecationMessage($this, $value))) {
+                        throw new InvalidOptionsException(sprintf('Invalid type for deprecation message, expected string but got "%s", returns an empty string to ignore.', \gettype($deprecationMessage)));
```
Comment (xabbuh) [inline src/Symfony/Component/OptionsResolver/OptionsResolver.php]: return an empty string

File: src/Symfony/Component/OptionsResolver/OptionsResolver.php [hunk 2]
```diff
@@ -360,7 +360,7 @@ public function getDefinedOptions()
      * Instead of passing the message, you may also pass a closure with the
      * following signature:
      *
-     *     function ($value) {
+     *     function (Options $options, $value) {
```
Comment (xabbuh) [inline src/Symfony/Component/OptionsResolver/OptionsResolver.php]: We could add the `string` return type hint here.

**Review by xabbuh — APPROVED** [review]
just two minor suggestions

**Comment by yceruto** [comment]
@xabbuh comments addressed. 

I'm preparing the doc PR right now.

**Review by javiereguiluz — APPROVED** [review]
Useful feature, nicely developed and fully documented. @yceruto thanks a lot for doing things so nice!

**Review by fabpot — APPROVED** [review]

**Pull request merged**